bus-mapping/src/circuit_input_builder.rs

```diff
@@ -1524,9 +1524,17 @@ impl<'a> CircuitInputBuilder {
         eth_block: &EthBlock,
         geth_traces: &[eth_types::GethExecTrace],
     ) -> Result<(), Error> {
+        // accumulates gas across all txs in the block
+        let mut cumulative_gas_used = HashMap::new();
```

Comment:
Here's a better idea: Introduce the field `cumulative_gas_used` in `CircuitInputBuilder::Transaction` which is initialized at 0.  Then in `gen_end_tx_ops` you just need to read the previous value with `state.block.txs[state.tx_ctx.id() - 1].cumulative_gas_used` and assign the new value to `state.tx.cumulative_gas_used`.

Finally you also add `cumulative_gas_used` in `evm_circuit::witness::Transaction`, and then in `EndTxGadget::assign_exec_step` you can just read the value from `block.txs[i].cumulative_gas_used`.

With this change you don't need to read that value in the assign function from the `block.rws`, and thus you can get rid of `TxReceiptFieldTag::iterator` and `TxReceiptFieldTag::amount`.

What do you think?

Reply:
@ed255 thanks for the idea!!  I understand `CircuitInputBuilder::Transaction` represents the concept of transaction in ethereum code,  existing fields are its intrinsic property, like `nonce` `gas` `from` `to`  etc.  but the `cumulative_gas_used` is an statistic value not only relate to current tx but also relying on previous txs.  moving it into `Transaction` may violates the raw structure purpose.   
Any thinking ?

Reply:
The idea of the `Block` and `Transaction` structs found in `CircuitInputBuilder` is to keep all the block and transaction values that the circuits will need as input, and that extends the intrinsic fields.
For example, in `Block` we have `history_hashes` and `container`, in `Transaction` we have `calls` and `steps`.
For structs that match the ethereum data structure we have `eth_types::{Block, Transaction}`.

Since there's a 1 to 1 match between transactions and transaction receipt, I think adding `cumulative_gas_used` to `CircuitInputBuilder::Transaction` (as a field corresponding to the receipt of that transaction) fits; and I believe this simplifies the handling of the `cumulative_gas_used` at `EndTx`.

Reply:
agree that it simplifies the handling of the cumulative_gas_used at EndTx.  `call` and `steps` can be derived from single tx,  they can be seen implicitly properties  while `cumulative_gas_used` can not refer by single tx though it is 1:1 mapping.   it can be done in practice, but somewhat different of concept .  let's hear more ideas , i.e. @han0110 what do you think about this?

Reply:
I have a try locally, by adding `cumulative_gas_used` field into  `Transaction`,  we need do following:
1.  handle `CircuitInputBuilder::Transaction` initialization with new field
2.  set `cumulative_gas_used` value in `handle_tx` 
3.  add `cumulative_gas_used`  field as well in `evm_circuit::witness::Transaction ` 
4.  update `evm_circuit::witness::tx_convert` to propagate new field value   

finally , we reach here  :  
```
 // let current_cumulative_gas_used: u64 = if tx.id == 1 {
        //     0
        // } else {
        //     let rw = &block.rws[(
        //         RwTableTag::TxReceipt,
        //         (tx.id - 1) * TxReceiptFieldTag::amount() - 1,
        //     )];
        //     rw.receipt_value()
        // };
// below is using cumulative_gas_used as transaction field way.
let current_cumulative_gas_used: u64 = if tx.id == 1 {
                0
            } else {
                block.txs[tx.id - 2].cumulative_gas_used
        };
```.
it seems to me bring more complexity just for this little place.

Reply:
> agree that it simplifies the handling of the cumulative_gas_used at EndTx. `calls` and `steps` can be derived from single tx, they can be seen implicitly properties of tx while `cumulative_gas_used` can not refer by single tx though it is 1:1 mapping. it can be done in practice, but somewhat different of concept . let's hear more ideas , i.e. @han0110 what do you think about this?

I understand your concern about `cumulative_gas_used` not being a value derived from a single tx.  In such case, I would say that another option (more code but more correct) is to define a struct `Receipt { cumulative_gas_used }` and add a new field to `CircuitInputBuilder::Block` `receipts: Vec<Receipt>`.  Then in `handle_tx`, an empty `Receipt` is appended to `block.receipts` and it is later set in the `gen_associated_ops` for `EndTx`.

Reply:
> I have a try locally, by adding `cumulative_gas_used` field into `Transaction`, we need do following:
> 
>     1. handle `CircuitInputBuilder::Transaction` initialization with new field
> 
>     2. set `cumulative_gas_used` value in `handle_tx`
> 
>     3. add `cumulative_gas_used`  field as well in `evm_circuit::witness::Transaction `
> 
>     4. update `evm_circuit::witness::tx_convert` to propagate new field value
> 
> 
> finally , we reach here in `end_tx` :
> 
> ```
>  // let current_cumulative_gas_used: u64 = if tx.id == 1 {
>         //     0
>         // } else {
>         //     let rw = &block.rws[(
>         //         RwTableTag::TxReceipt,
>         //         (tx.id - 1) * TxReceiptFieldTag::amount() - 1,
>         //     )];
>         //     rw.receipt_value()
>         // };
> // below is using cumulative_gas_used as transaction field way.
> let current_cumulative_gas_used: u64 = if tx.id == 1 {
>                 0
>             } else {
>                 block.txs[tx.id - 2].cumulative_gas_used
>         };
> ```
> 
> it seems to me bring more complexity just for this little place.

Bear in mind that by following this approach you can remove the implementation of `TxReceiptFieldTag::iterator` and `TxReceiptFieldTag::amount`; and with that I don't see it being more complex than the current implementation in the PR (I find it more easy to follow).

Nevertheless your current implementation is correct :)  My comment was a proposal: it's OK if after discussing it we don't reach an agreement; and in that case we can proceed with your original implementation

Reply:
I'm marking this conversation as resolved as this was a proposal and we didn't reach consensus about it :P